### ISSUE TITLE
FR: custom domains for magic DNS

### ISSUE THREAD

**Issue by fairplaymvv** [tailscale/tailscale#4221]
### What are you trying to do?

As a user, I'd like to point to a custom domain instead of `<domain>.beta.tailscale.net`, e.g. `foobar.network`, where `foobar.network` is in my or my org's possession.

### What is the impact of not solving this?

I use the more cludgy domain than the one and I'm going to have to probably figure out my ssh config strategy all over again ¯\_(ツ)_/¯

**Comment by mayakacz** [comment]
Duplicate of https://github.com/tailscale/tailscale/issues/1543

**Comment by bradfitz** [comment]
I don't think those are quite the same, @mayakacz. #1543 is about adding custom records. This is about BYO MagicDNS domain suffix I think?

**Comment by joshuataylor** [comment]
Sorry for the necro, but I would love this as well! Tailscale MagicDNS is 🔥 and actually 🪄, but would love to BYO domain for this part:
https://tailscale.com/kb/1081/magicdns/#full-domain-names-vs-machine-names

So we replace domain + suffix with say:
`monitoring.example.com`, where example.com would be the host. Probably wise to have third level (`monitoring.foobar.example.com`, similar to existing?).

My custom domain is shorter, it's not a deal breaker, more of a "this would be nice to have". Love the random name generator tailscale provides. :-)

**Comment by dsymonds** [comment]
Yes, could this issue be reopened? With MagicDNS going GA today this would make it even more attractive.

**Comment by DentonGentry** [comment]
In retrospect, not a duplicate of https://github.com/tailscale/tailscale/issues/1543

**Comment by viperfx** [comment]
Would really love to see this feature. Have a custom domain and would love to use it with sub sub domain support with https for my NAS box.

**Comment by talwat** [comment]
I would also like this quite a lot.

**Comment by hazelement** [comment]
+1 on this. I'd like to use the same domain while I'm on tailscale vs on my physical network.

**Comment by MSRSPSR** [comment]
@panekj Mind saying why you're downvoting this?

**Comment by DentonGentry** [comment]
We understand the interest and we use the emoji reactions on the original issue as part of prioritization of what to work on. To express interest, please add a reaction.

I'm going to close the issue for further discussion, as comments expressing concurrence don't get counted so easily (we cannot generate a report of counts from GitHub, for example).

**Comment by chrishas35** [comment]
To add to the request, when enabling this, I would like to be able to delegate a portion of my existing domain to Tailscale DNS for management.  For example, sending *.ts.example.com to 100.100.100.100 (this might need to be a named record, so tailscale may need to publish a public A record for 100.100.100.100?)

**Comment by frellus** [comment]
Would add to this, as a new potential user to tailscale, this would greatly help us on-board an existing non-tailscale network9s)/domain into tailscale, as with a large organization of nodes and users I need to move things in pieces. Having a *.ts.example.com, or even a completely hidden domain name matching my network name like *.secured which I can re-direct to the magicDNS from my traditional DNS servers offers a way I can move server/service-by-service into the "secure" domain in a very clean way.

Am surprised this isn't higher priority - ZeroTier offers this and I feel with large organization this is tablestakes - thanks!

**Comment by lloydjatkinson** [comment]
I also wanted to add that I think neither the randomly generated default name **or** the selectable “fun” names are great for branding/name recognition/basic security especially for non-technical users. Both types of name could be perceived to be suspicious. A well known name, such as organisation name etc would be easier to remember.

**Comment by evilhamsterman** [comment]
At least for paying companies it would be good to support company.ts.net as a start.

**Comment by chrishas35** [comment]
While not the first-party solution being asked for here, I did recently see this blog post from @willnorris that sets this up with an external solution

https://willnorris.com/2023/tailscale-custom-domain/

**Comment by cpressland** [comment]
One thing we'd like some _conceptual feedback_ on from Tailscale is what implications building a feature like this would have on customers that have already enabled HTTPS Certificates.

I'm keen to enable HTTPS Certificates so we can use Kubernetes Ingress via the Tailscale Kube Operator, however, if this feature ends up landing `thing.bink.ts.net` would be far preferable to `thing.fun-name.ts.net`. The documentation clearly states that once HTTPS Certificates are enabled the tailnet name cannot be changed. So either exceptions would need to be granted by the support team on a request by request basis or this policy would need to be relaxed.

@DentonGentry - could you offer any feedback on if I should just stop worrying and enable HTTPS Certificates or if I'm potentially better to hold off and wait for this feature as this policy would not be relaxed?

**Comment by DentonGentry** [comment]
> could you offer any feedback on if I should just stop worrying and enable HTTPS Certificates or if I'm potentially
> better to hold off and wait for this feature as this policy would not be relaxed?

This feature is not imminent. Holding off on use of HTTPS certificates would mean holding off for an unknown amount of time. If HTTPS certificates as they exist today solve a problem, I'd suggest using them.

I cannot speak to exactly what the policy regarding would be regarding tailnets which began using HTTP certificates prior to the existence of a feature such as this. The "exceptions would need to be granted by the support team" path would exist.

**Comment by xylobol** [comment]
Adding the note that I'd like this feature even if it means using a custom ACME-compatible CA.

**Comment by Investigamer** [comment]
+1 for this request, really hope it gains traction. I wouldn't even mind if it was a paid plan feature, it would just make my life that much simpler.

**Comment by jwpauler** [comment]
+1. In case you hadn't noticed, the "fun tailnet names" are not fun. I would never tell a soul that I was using woodpecker-altered.ts.net or chimp-edmontosaurus.ts.net. For now, I'll stick with tailxxxxx.ts.net. I'll still get laughed at, for sure, but they won't laugh as hard as someone using neurotrichini-macaroni.ts.net

**Comment by alext-extracellular** [comment]
I don't know if it helps but I would also like to add my voice for being able to do <company_name>.ts.net.
I am an enterprise user, and sharing the fun name with colleagues and expecting them to click it is a game of chance.

**Comment by evilhamsterman** [comment]
> I don't know if it helps but I would also like to add my voice for being able to do <company_name>.ts.net.
> I am an enterprise user, and sharing the fun name with colleagues and expecting them to click it is a game of chance.

@alext-extracellular they can already do this for Enterprise users, contact your rep. There are some advisories to be aware of but they can do it any time

**Comment by casperfrx** [comment]
Seemed like a given that such cases would be supported on the enterprise level because binding the functionality to `<fun-name>.ts.net` is a choice after all. The `tailscale cert` functionality should be expandable in the code to add private DNS-challenge validated domains, that way I wouldn't need to run a redundant reverse proxy to the same ports either `tailscale funnel` or `tailscale serve` occupies and alleviates port conflicts/unnecessary workarounds. But alas. Still love the product.

PS: maybe the upcoming change towards short-lived TLS certs may speed it up a little eventually where Tailscale has to take security in consideration

**Comment by anthonysomerset** [comment]
> I don't know if it helps but I would also like to add my voice for being able to do <company_name>.ts.net. I am an enterprise user, and sharing the fun name with colleagues and expecting them to click it is a game of chance.

even as a non enterprise user (at this time - attempting to evaluate fit) i would love for my home lab to be able to chose my own <arbitrarystring>.ts.net rather than randomly generated fun ones

**Comment by evilhamsterman** [comment]
Fully custom domains would be great, I can see the issues with custom ts.net domains. Maybe a middle ground, since custom ts.net domains are similar to regular domains which have to be purchased and renewed regularly, how about offering a custom ts.net domain for a nominal yearly fee. I'd be fine ponying up $10-15/year for something like evilhamsters.ts.net, it would also be an additional, albeit small, revenue stream. Those who don't care can use random of "fun" domains, those wanting more pay a few bucks to support the company and get a pick. If they stop paying you can reclaim the name.

**Comment by Sheumack** [comment]
Also voting for this issue. 

Echoing points made above - 

- tsNNNNNN.ts.net fqdn addresses aren't bad. They'll do at a push. But for anything customer facing, a branded URL looks much more professional.  

- The "roll of the dice" words aren't really great options for business: MyProfessionalService.rancid-dinosaur.ts.net doesn't look great in front of customers. For a home network they're possibly suitable.
So while obviously it would be great to have this for everybody, if it needs to go onto the $6 a month business plan, fair enough.

- Yes, it's relatively straight forward to stand up and maintain your own DNS server for the local domain (CoreDNS, dnsmasq, etc), but that could be equally be said about other core Tailscale features - the whole point of buying a service that handles this stuff is so you don't have to. 

- Fully appreciate a land-grab for "myNameHere.ts.net" addresses is unnecessary and unwanted dispute resolution overhead. There is not authoritative international trademark or business name register that ensures global uniqueness, and getting in the middle of squabbles over naming rights is a distraction the company doesn't need.

However, domain names are globally unique by design, so if somebody can demonstrate they own domain, then there is no conflict of interest to resolve (pun intended). that could be ts_subdomain.primarydomain.com or a dedicated 2nd level domain like primarydomain_lab.com ... whatever people want to register. Mechanisms already exist to permit this sort of validation, for example adding a TXT record to DNS. Hosting and email providers do this as matter of course.

(If somebody "accidentally" puts their primary public domain into MagicDNS and completely destroys their entire internal routing to their public sites... well.. they'll only do it once. You can't control everything. Put up a warning and leave it to personal responsibility).

**Comment by Illustar0** [comment]
This feature request has been open for four years now, and it is currently the second most upvoted open feature request.

Honestly, I do not think this feature is technically very difficult to implement. After all, Tailscale already allows users to choose a “fun” tailnet name. So the only reasonable explanation I can think of is that custom MagicDNS domains are being intentionally kept as an Enterprise-only feature behind a paywall — and the fact that this feature is indeed available on the Enterprise plan seems to confirm that.

**Comment by willnorris** [comment]
> and the fact that this feature is indeed available on the Enterprise plan seems to confirm that

I'm not sure which feature you're referring to here, but we don't have the ability to use a custom domain for device FQDNs, enterprise plan or otherwise.  It is definitely a highly requested feature, and one I want personally myself.  There are a number of reasons it hasn't been built yet, but none of them great... mostly delays due to silly internal technical challenges, and then just the simple fact that other things have taken priority.

**Comment by evilhamsterman** [comment]
@Illustar0 I also agree this would be a nice feature. Though I think you are confusing custom domains with custom Tailnet names. Being on an Enterprise plan at work I can confirm that we were able to request a custom Tailnet name of `<company>.ts.net` but that is different from being able to use an entirely custom FQDN like `ts.<company>.com`.

Both are slightly related but really two separate things.

**Comment by Illustar0** [comment]
Oh yes, sorry — you’re right. I confused custom Tailnet names with custom MagicDNS domains.

The reason I made that assumption is that #8240, “Custom Tailnet Name”, was marked as a duplicate of #4221, “custom domains for MagicDNS”, so I thought they were being treated as the same or closely related feature request.

Thanks for clarifying.

**Comment by Sheumack** [comment]
I just want to clarify why I conflated Domain names with Custom Tailscale .ts.net names... which is that it removed the administrative burden from Tailscale to deal with the noise that will inevitably come from conflicts over who has the right to a particular ts name. 

We already know the product can provide this functionality, but I can imagine that suddenly thousands of users trying to snarf their favourite word would trigger e a flood of annoying tickets that's distracting from the real work of building tailscale.

If they were to only permit globally registered domains to be used as the prefix to .ts.net, then the global uniqueness and ownership is already enforced.  That would required a mechanism to periodically confirm rightful ownership, but that mechanism is established using TEXT records in the registrar's DNS.

Anyway... my workaround in the end was an internal custom dns:
- Spin up a "free tier" ubuntu VM in google cloud (it actually costs me about US$0.60 a month, but I'll cope with that.
- Install tailscale and dnsmasq
- Make dnsmasq authorative for ts._domain_  (or lab._domain_ or whatever)
- Create a cronjob that periodically runs a tailscale status, takes the first two columns, appends the .ts.<domain> to the end of each line, dumps it into a temp file, runs a diff against /etc/dnsmasq._domain_.hosts,  If there are changes, rotate the file to a _datetime, move the temp file to the new hosts file, reload dnsmasq, else dump the tmpfile.

It's not perfect, and definitely not pretty enough to share (or deploy in a real enterprise), but it's functional.  So it can be done.

Getting the Split DNS configuration was a little tricky, but got there in the end so there were no circular dependencies, and the hosts outside the specific subdomain resolved correctly. 

This is a workaround, and does not close the request as it does require quite a bit of faffing about to get it working, and 90% of the functionality this provides would have been solved by a custom subdomain under .ts.net.

**Comment by willnorris** [comment]
> Anyway... my workaround in the end was an internal custom dns:

If you didn't see my blog post that someone linked above, it might be worth a look... https://willnorris.com/2023/tailscale-custom-domain/.  It sounds like it achieves that same thing you've got (I think), but with fewer moving parts.  It's just a coredns server with a Tailscale plugin that resolves queries based on a live view of the Tailscale netmap (so no cronjob to update records).

The challenging parts to fully handle this feature were domain verification, which we just recently launched, and then changing our DNS infrastructure a bit.  Right now, we write public DNS records if you use Tailscale Funnel or issue certificates (we publish records for ACME DNS-01 challenges).  Right now, because all tailnets are subdomains of `ts.net`, we can have a single zone file that we publish records into.  When customers are bringing their own domains, they need to delegate to our DNS servers, and then we need to publish individual zone files for each customer.  All very doable, but just a very different structure than what we have today, so it'll just take some work to restructure things.

The good news is, at least the last time I tested it a year or so ago, the Tailscale client handles the custom domain just fine.  It doesn't really care what domain you use for magicdns, which I believe is what also allows that to work fine with headscale.  We just need to make some changes in our control server for this to work, it likely won't require significant (or any?) client changes.
